doc/source/rllib-concepts.rst

```diff
@@ -581,12 +581,141 @@ Here is an example of creating a set of rollout workers and using them gather ex
+
+In code, this dataflow can be expressed as the following execution plan:
+
+.. code-block:: python
+
+    def execution_plan(workers: List[RolloutWorker], config: TrainerConfigDict):
+        # type: LocalIterator[SampleBatchType]
+        rollouts = ParallelRollouts(workers, mode="bulk_sync")
+
+        # type: LocalIterator[(SampleBatchType, List[LearnerStatsDict])]
+        train_op = rollouts \
+            .combine(ConcatBatches(
+                min_batch_size=config["train_batch_size"])) \
+            .for_each(TrainOneStep(workers))
+
+        # type: LocalIterator[ResultDict]
+        return StandardMetricsReporting(train_op, workers, config)
+
+
+As you can see, each step returns an *iterator* over objects. These iterators represent the infinite stream of data items that can be produced from the dataflow. Each operator (e.g., ``ConcatBatches``, ``TrainOneStep``), executes an operation over each item and returns a transformed item (e.g., concatenated batches, learner stats from training). Finally, some operators such as TrainOneStep have the *side-effect* of updating the rollout worker weights (that's why ``TrainOneStep`` takes the list of worker actors ``workers`` as an argument).
+
+Understanding and Debugging Execution Plans
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Execution plans are based on Ray `parallel iterators <iter.html>`__ and can be inspected similarly. For example, suppose you wanted to print out the intermediate data items during training. This can be done by inserting a print function into the dataflow, e.g., for A2C:
+
+.. code-block:: python
+
+    def debug_print(item):
+        print("I saw", type(item))
+        return item
+
+    train_op = rollouts \
+        .combine(ConcatBatches(
+            min_batch_size=config["train_batch_size"])) \
+        .for_each(debug_print) \
+        .for_each(TrainOneStep(workers))
+
+You'll see output like this on the console:
+
+.. code-block:: bash
+
+    (pid=6555) I saw <class 'ray.rllib.policy.sample_batch.SampleBatch'>
+    (pid=6555) I saw <class 'ray.rllib.policy.sample_batch.SampleBatch'>
+    (pid=6555) I saw <class 'ray.rllib.policy.sample_batch.SampleBatch'>
+    (pid=6555) I saw <class 'ray.rllib.policy.sample_batch.SampleBatch'>
+
+It is important to understand that the iterators of an execution plan are evaluated *lazily*. This means that no computation happens until the `trainer <#trainers>`__ tries to read the next item from the iterator (i.e., get the next training result for a ``Trainer.train()`` call).
+
+Execution Plan Concepts
+~~~~~~~~~~~~~~~~~~~~~~~
+
+RLlib provides a library of operators `(GitHub link) <https://github.com/ray-project/ray/tree/master/rllib/execution>`__ that can be used in execution plans. You can of course write your own operators (which are just normal Python functions). A few categories of operators are summarized below:
```

Comment:
Let's remind the user again here what an op actually (besides a function), like what the inputs and outputs to that function always are (e.g. point to above debug print op, or show another simple custom one).